File: PetBookUI/core/typing.cs
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using PetBook.core;
namespace PetBook.typing
{
    public class IntStringPair
    {
        public IntStringPair(int id, string name)
        {
            this.i = id;
            this.s = name;
        }

        public int i { get; set; }
        public string s { get; set; }
    }

    unsafe public class Effect
    {
        public int id;
        public int cnt;
        public string desc;
        public string args;
        public Effect(CEffect* o)
        {
            this.id = o->id;
            this.cnt = o->cnt;
            this.desc = Marshal.PtrToStringUTF8((IntPtr)o->description)!;
            this.args = Marshal.PtrToStringUTF8((IntPtr)o->args)!;
        }
    }

    unsafe public class Skill
    {
        public int id { get; set; }
        public string name { get; set; }

        public int category { get; set; }
        public int types { get; set; }
        public String types_string
        {
            get { return MainWindow.mapping[types].Item1; }
        }
        public int power { get; set; }
        public int pp { get; set; }
        public int accuracy { get; set; }
        public int critical_rate { get; set; }

        public List<Effect> effects { get; set; }

        public string info { get; set; }
        public int lv { get; set; }
        public int extra { get; set; }
        public Skill(CSkill* cur)
        {
            Int32 id = cur->id;
            string name = Marshal.PtrToStringUTF8((IntPtr)cur->name)!;

            Int32 category = cur->category;
            Int32 types = cur->types;
            Int32 power = cur->power;
            Int32 pp = cur->pp;
            Int32 accuracy = cur->accuracy;
            Int32 critical_rate = cur->critical_rate;

            List<Effect> effects = new List<Effect>();
            CArray* ptr = (CArray*)cur->effect;
            CEffect* o = (CEffect*)ptr->ptr;
            for (int j = 0; j < ptr->len; j += 1)
            {
                effects.Add(new Effect(o));
                o += 1;
            }

            string info = Marshal.PtrToStringUTF8((IntPtr)cur->info)!;
            Int32 lv = cur->lv;
            Int32 extra = cur->extra;
            this.id = id;
            this.name = name;
            this.category = category;
            this.types = types;
            this.power = power;
            this.accuracy = accuracy;
            this.pp = pp;
            this.critical_rate = critical_rate;
            this.effects = effects;
            this.info = info;
            this.lv = lv;
            this.extra = extra;
        }
    }

    public class Stringx3
    {
        public string s1 { get; set; }
        public string s2 { get; set; }
        public string s3 { get; set; }

        public Stringx3(string s1, string s2, string s3)
        {
            this.s1 = s1;
            this.s2 = s2;
            this.s3 = s3;
        }

    }

    public struct MintMark
    {
        public int id { get; set; }
        public string name { get; set; }

        public int hp { get; set; }
        public int atk { get; set; }
        public int def { get; set; }
        public int spatk { get; set; }
        public int spdef { get; set; }
        public int spd { get; set; }
        public int sum { get; set; }
        public unsafe MintMark(CMintMark* x)
        {
            this.id = x->id;
            this.name = Marshal.PtrToStringUTF8((IntPtr)x->name)!;
            this.hp = x->hp;
            this.spd = x->spd;
            this.atk = x->atk;
            this.def = x->def;
            this.spatk = x->spatk;
            this.spdef = x->spdef;
            this.sum = x->hp + x->spd + x->atk + x->def + x->spatk + x->spdef;
        }
    }

}

File: src/typing.rs
use std::ffi::{c_void, CString};
use std::mem::ManuallyDrop;
use std::os::raw::c_char;

#[repr(C, align(4))]
pub struct Array {
    pub ptr: *const c_void,
    pub len: u32,
    pub cap: u32,
}

impl Array {
    pub fn new<T>(src: Vec<T>) -> Self {
        let src = ManuallyDrop::new(src);
        Self {
            ptr: src.as_ptr() as *const c_void,
            len: src.len() as u32,
            cap: src.capacity() as u32,
        }
    }
}

#[repr(C)]
pub struct BaseDetail {
    pub id: u32,
    pub name: *const c_char,
}

impl BaseDetail {
    pub fn new(id: u32, name: String) -> Self {
        let str = CString::new(name).unwrap();
        let ptr = str.into_raw();
        Self { id, name: ptr }
    }
}

#[repr(C)]
pub struct PetFullDetail {
    pub base: BaseDetail,
    pub hp: u32,
    pub atk: u32,
    pub def: u32,
    pub spatk: u32,
    pub spdef: u32,
    pub spd: u32,

    pub types: u32,
    pub gender: u32,
    pub evolving_lv: u32,
}

#[repr(C)]
pub struct Effect {
    pub id: i32,
    pub cnt: i32,
    pub args: *const c_char,
    pub description: *const c_char,
}

#[derive(Debug)]
#[repr(C)]
pub struct Skill {
    pub id: i32,
    pub name: *const c_char,
    pub category: i32,
    pub types: i32,
    pub power: i32,
    pub pp: i32,
    pub accuracy: i32,
    pub critical_rate: i32,
    pub effect: *const c_void,
    pub info: *const c_char,
    pub lv: i32,
    pub extra: i32,
}

#[derive(Debug)]
#[repr(C)]
pub struct SkillType {
    pub id: i32,
    pub name: *const c_char,
    pub is_double: u8,
}

#[derive(Debug)]
#[repr(C)]
pub struct MintMark {
    pub id: i32,
    pub name: *const c_char,
    pub atk: i32,
    pub def: i32,
    pub spatk: i32,
    pub spdef: i32,
    pub spd: i32,
    pub hp: i32,
}

pub enum Status {
    UpdateSkillIcon = -3,
    Idle = -2,
    UpdateDatabase = -1,
}

#[derive(Debug)]
#[repr(C)]
pub struct Buff {
    pub id: i32,
    pub effect_id: i32,
    pub tips: *const c_char,
    pub come: *const c_char,
    pub url: *const c_char,
}

pub enum OrderBy {
    Id = 1,
    IdDesc = 2,
    Hp = 4,
    Atk = 8,
    Def = 16,
    Spatk = 32,
    Spdef = 64,
    Spd = 128,
    Sum = 256,
}

impl OrderBy {
    pub fn new(val: i32) -> Self {
        match val {
            1 => OrderBy::Id,
            2 => OrderBy::IdDesc,
            4 => OrderBy::Hp,
            8 => OrderBy::Atk,
            16 => OrderBy::Def,
            32 => OrderBy::Spatk,
            64 => OrderBy::Spdef,
            128 => OrderBy::Spd,
            256 => OrderBy::Sum,

            _ => {
                panic!("bad order")
            }
        }
    }
    pub fn to_str(&self) -> String {
        let o = match *self {
            OrderBy::Hp => "hp desc",
            OrderBy::Atk => "atk desc",
            OrderBy::Def => "def desc",
            OrderBy::Spatk => "spatk desc",
            OrderBy::Spdef => "spdef desc",
            OrderBy::Spd => "spd desc",
            OrderBy::Sum => "hp+atk+def+spatk+spdef+spd desc",
            OrderBy::Id => "id",
            OrderBy::IdDesc => "id desc",
        };
        format!("order by {}", o)
    }
}
